File: backend/app/routers/graph.py
from fastapi import APIRouter, HTTPException

from app.core.logging import get_logger
from app.schemas.graph import EntityChunk, GraphResponse, RebuildSummary
from app.services import graph_extractor, graph_store, vectorstore
from app.services.vectorstore import VectorStoreError

logger = get_logger(__name__)
router = APIRouter(prefix="/api/graph", tags=["graph"])


@router.get("", response_model=GraphResponse)
async def get_graph(doc_id: str | None = None) -> GraphResponse:
    """Renvoie le graphe (optionnellement restreint à un document)."""
    data = await graph_store.get_graph(doc_id)
    return GraphResponse(**data)


@router.post("/rebuild", response_model=RebuildSummary)
async def rebuild_graph() -> RebuildSummary:
    """(Re)construit le graphe pour TOUS les documents indexés (séquentiel)."""
    try:
        docs = await vectorstore.list_documents()
    except VectorStoreError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc

    logger.info("Graphe : reconstruction complète sur %d document(s)…", len(docs))
    await graph_store.clear_all()

    total_chunks = 0
    for n, doc in enumerate(docs, start=1):
        doc_id, doc_name = doc["doc_id"], doc["name"]
        try:
            chunks = await vectorstore.get_all_chunks({"doc_id": doc_id})
        except VectorStoreError as exc:
            raise HTTPException(status_code=503, detail=str(exc)) from exc
        logger.info("Graphe : document %d/%d — '%s'", n, len(docs), doc_name)
        result = await graph_extractor.extract_doc(doc_id, doc_name, chunks)
        total_chunks += result["chunks"]

    stats = await graph_store.stats()
    logger.info(
        "Graphe : reconstruction terminée (%d nœuds, %d arêtes).",
        stats["nodes"],
        stats["edges"],
    )
    return RebuildSummary(
        documents=len(docs),
        nodes=stats["nodes"],
        edges=stats["edges"],
        chunks=total_chunks,
    )


@router.get("/entity/{node_id}/chunks", response_model=list[EntityChunk])
async def get_entity_chunks(node_id: str) -> list[EntityChunk]:
    """Liste les chunks qui mentionnent une entité (avec texte + nom de doc)."""
    rows = await graph_store.get_entity_chunks(node_id)
    if not rows:
        return []

    try:
        docs = await vectorstore.list_documents()
        names = {d["doc_id"]: d["name"] for d in docs}
        # Texte des chunks, récupéré par document depuis Chroma.
        texts: dict[str, str] = {}
        for did in {r["doc_id"] for r in rows}:
            for c in await vectorstore.get_all_chunks({"doc_id": did}):
                texts[c["id"]] = c["text"]
    except VectorStoreError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc

    return [
        EntityChunk(
            doc_id=r["doc_id"],
            doc_name=names.get(r["doc_id"], ""),
            page=r["page"],
            chunk_id=r["chunk_id"],
            text=texts.get(r["chunk_id"], ""),
        )
        for r in rows
    ]

File: frontend/src/lib/graph.ts
import { API_BASE } from "./api";

export type EntityType =
  | "person"
  | "org"
  | "place"
  | "date"
  | "concept"
  | "other";

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  count: number;
}

export interface GraphEdge {
  source: string;
  target: string;
  label: string;
  weight: number;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface EntityChunk {
  doc_id: string;
  doc_name: string;
  page: number;
  chunk_id: string;
  text: string;
}

export interface RebuildSummary {
  documents: number;
  nodes: number;
  edges: number;
  chunks: number;
}

async function errText(res: Response): Promise<string> {
  try {
    const body = await res.json();
    if (typeof body?.detail === "string") return body.detail;
  } catch {
    /* corps non-JSON */
  }
  return `Erreur ${res.status}`;
}

/** Récupère le graphe (optionnellement restreint à un document). */
export async function fetchGraph(docId?: string | null): Promise<GraphData> {
  const url = new URL(`${API_BASE}/api/graph`);
  if (docId) url.searchParams.set("doc_id", docId);
  const res = await fetch(url.toString());
  if (!res.ok) throw new Error(await errText(res));
  return res.json();
}

/** Chunks qui mentionnent une entité (avec texte + nom de document). */
export async function fetchEntityChunks(nodeId: string): Promise<EntityChunk[]> {
  const res = await fetch(
    `${API_BASE}/api/graph/entity/${encodeURIComponent(nodeId)}/chunks`,
  );
  if (!res.ok) throw new Error(await errText(res));
  return res.json();
}

/** (Re)construit le graphe pour tous les documents indexés. Lent sur CPU. */
export async function rebuildGraph(): Promise<RebuildSummary> {
  const res = await fetch(`${API_BASE}/api/graph/rebuild`, { method: "POST" });
  if (!res.ok) throw new Error(await errText(res));
  return res.json();
}
